infer/internal/ende/ende.go

```diff
@@ -0,0 +1,352 @@
+		if c.emptyAction == isNil {
+			c.emptyAction = o.emptyAction
+		}
+
+		e.changes = e.changes[:len(e.changes)-1]
+	}
+	e.changes = append(e.changes, c)
+}
+
+func (e *ende) walk(
+	v resource.PropertyValue, path resource.PropertyPath, typ reflect.Type,
+	alignTypes bool,
+) resource.PropertyValue {
+	if typ == nil {
+		// We can't align types when we don't have type info
+		alignTypes = false
+	} else {
+		for typ.Kind() == reflect.Pointer {
+			typ = typ.Elem()
+		}
+	}
+
+	switch {
+	case v.IsSecret():
+		// To allow full fidelity reconstructing maps, we extract nested secrets
+		// first. We then extract the top level secret. We need this ordering to
+		// re-embed nested secrets.
+		el := e.walk(v.SecretValue().Element, path, typ, alignTypes)
+		e.mark(change{path: path, secret: true})
+		return el
+	case v.IsComputed():
+		el := e.walk(v.Input().Element, path, typ, true)
+		e.mark(change{path: path, computed: true})
+		return el
+	case v.IsOutput():
+		output := v.OutputValue()
+		el := e.walk(output.Element, path, typ, !output.Known)
+		e.mark(change{
+			path:        path,
+			computed:    !output.Known,
+			secret:      output.Secret,
+			forceOutput: true,
+		})
+
+		return el
+	}
+
+	var elemType reflect.Type
+	if typ != nil {
+		switch typ.Kind() {
+		case reflect.Slice, reflect.Array, reflect.Map:
+			elemType = typ.Elem()
+		}
+	}
+
+	if !alignTypes {
+		switch {
+		case v.IsArray():
+			return e.walkArray(v, path, elemType, alignTypes)
+		case v.IsObject():
+			// We need to walk structs in a strongly typed way, so we omit
+			// them here.
+			if typ == nil || typ.Kind() != reflect.Struct {
+				return e.walkMap(v, path, elemType, alignTypes)
+			}
+		// This is a scalar value, so we can return it as is.
+		default:
+			return v
+		}
+	}
+
+	contract.Assertf(!IsComputed(v), "failed to strip computed")
+	contract.Assertf(!IsSecret(v), "failed to strip secrets")
+	contract.Assertf(!v.IsOutput(), "failed to strip outputs")
+
+	switch typ.Kind() {
+	case reflect.Array, reflect.Slice:
+		return e.walkArray(v, path, elemType, alignTypes)
+	case reflect.Map:
+		return e.walkMap(v, path, elemType, alignTypes)
+	case reflect.Struct:
+		result := resource.PropertyMap{}
+		if v.IsObject() {
+			result = v.ObjectValue().Copy()
+		}
+		for _, field := range reflect.VisibleFields(typ) {
+			tag, err := introspect.ParseTag(field)
+			if err != nil || tag.Internal {
+				continue
+			}
+			pName := resource.PropertyKey(tag.Name)
+			path := append(path, tag.Name)
+			if v, ok := result[pName]; ok {
+				result[pName] = e.walk(v, path, field.Type, alignTypes)
+			} else {
+				if tag.Optional || !alignTypes {
+					continue
+				}
+				// Create a new unknown output, which we will then type
+				result[pName] = e.walk(resource.NewNullProperty(),
+					path, field.Type, true)
+			}
+		}
+		return resource.NewObjectProperty(result)
+	case reflect.String:
+		if v.IsString() {
+			return v
+		}
+		return resource.NewStringProperty("")
```

Comment:
Why do we default to returning nullish values if the type is wrong? It seems like it might be better to error if the actual type doesn't fit the expected type unless there's a reason we need less strict mapping?

Reply:
The engine sends us invalidly typed values behind unknowns. `output<bool, known:false>` will get sent as `resource.PropertyValue{V: resource.Output{Element: resource.PropertyValue{V: ""}, Known: true}}`. We need to do the type correction so the mapper doesn't error.